spec/unit/puppet_pal_2pec.rb

```diff
@@ -233,28 +233,100 @@
+          ctx.evaluate_script_string('a::afunc()')
+        end
+        expect(result).to eq("a::afunc value")
       end
-      expect(result).to eq("a::aplan value")
+
+      it 'a given "modulepath" overrides the default' do
+        expect do
+          result = Puppet::Pal.in_environment('pal_env', env_dir: testing_env_dir, modulepath: [], facts: node_facts) do |ctx|
+            ctx.evaluate_script_string('a::afunc()')
+          end
+        end.to raise_error(/Unknown function: 'a::afunc'/)
+      end
+
+      it 'a plan in a module can be called with run_plan' do
+        result = Puppet::Pal.in_environment('pal_env', env_dir: testing_env_dir, facts: node_facts) do |ctx|
+          ctx.evaluate_script_string('run_plan("a::aplan")')
+        end
+        expect(result).to eq("a::aplan value")
+      end
+
+      it 'errors in a meaningful way when a non existing env name is given' do
+        testing_env_dir # creates the structure
+        expect do
+          Puppet::Pal.in_environment('blah_env', env_dir: testing_env_dir.chop, facts: node_facts) do |ctx|
+            ctx.evaluate_script_string('run_plan("a::aplan")')
+          end
+        end.to raise_error(/The environment directory '.*' does not exist/)
+      end
+
     end
 
-    it 'sets the facts' do
+    context 'configured as existing given envpath such that' do
+      it 'modules in it are available from its "modules" directory' do
+        testing_env_dir # creates the structure
+        result = Puppet::Pal.in_environment('pal_env', envpath: environments_dir, facts: node_facts) do |ctx|
+          ctx.evaluate_script_string('a::afunc()')
+        end
+        expect(result).to eq("a::afunc value")
+      end
+
+      it 'a given "modulepath" overrides the default' do
+        testing_env_dir # creates the structure
+        expect do
+          result = Puppet::Pal.in_environment('pal_env', envpath: environments_dir, modulepath: [], facts: node_facts) do |ctx|
+            ctx.evaluate_script_string('a::afunc()')
+          end
+        end.to raise_error(/Unknown function: 'a::afunc'/)
+      end
+
+      it 'a plan in a module can be called with run_plan' do
+        testing_env_dir # creates the structure
+        result = Puppet::Pal.in_environment('pal_env', envpath: environments_dir, facts: node_facts) do |ctx|
+          ctx.evaluate_script_string('run_plan("a::aplan")')
+        end
+        expect(result).to eq("a::aplan value")
+      end
+
+      it 'errors in a meaningful way when a non existing env name is given' do
+        testing_env_dir # creates the structure
+        expect do
+          Puppet::Pal.in_environment('blah_env', envpath: environments_dir, facts: node_facts) do |ctx|
+            ctx.evaluate_script_string('run_plan("a::aplan")')
+          end
+        end.to raise_error(/No directory found for the environment 'blah_env' on the path '.*'/)
+      end
+
+    end
```

Comment:
Needs test for when both envdir and envpath are given.

Reply:
Since this is meant to be API, should have tests for invalid arguments. modulepath containing instances of File or Pathname, envdir that is something other than string, etc.

Reply:
Added a bunch of tests.